File: src/context_tms.rs
use crate::network::{ Network };
use crate::premise::Premise;
use crate::context::Context;
use crate::util::CellID;
use std::collections::HashSet;
use std::collections::HashMap;
use std::rc::{Rc, Weak};
use std::cell::RefCell;
use std::fmt;
use std::fmt::Debug;


pub enum Action {
    AmbChoose(CellID)
}

#[derive(Clone)]
pub struct TruthManagementContext<T> {
    premise_outness: HashSet<T>,
    premise_nogoods: HashMap<T, Vec<(T, T)>>,
}

impl<T> fmt::Debug for TruthManagementContext<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Truth Management System")

        //TODO premises
    }
}

impl<T: Premise> TruthManagementContext<T> {
    pub fn new() -> Self {
        Self {
            premise_outness: HashSet::new(),
            premise_nogoods: HashMap::new()
        }
    }

    pub fn premise_in(&self, premise: &T) -> bool {
        !self.premise_outness.contains(premise)
    }

    //fn pairwise_union() {
        //let mut vec = Vec:new();

        //for 
    //}
}

impl<T: Premise> Context for TruthManagementContext<T> {
    type Action = Action;

    fn run_action(&mut self, action: Self::Action) {
        match action {
            Action::AmbChoose(cell_id) => {
                let true_premise : T = Premise::make_hypothetical(true, cell_id);
                let false_premise : T = Premise::make_hypothetical(false, cell_id);

                let true_no_goods = self.premise_nogoods.get(&true_premise);
                let false_no_goods = self.premise_nogoods.get(&false_premise);

                let reasons_agains_true = match true_no_goods {
                    Some(no_goods) => {
                        no_goods.iter().filter(|premises| {
                            true
                        }).collect()
                    },
                    None => Vec::new()
                };
            }
        }
    }

}

File: src/context_tms.zig
const std = @import("std");
const CellID = @import("util.zig").CellID;
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;

pub fn Premise(comptime T: type) type {
    return union(enum) {
        const Self = @This();

        HypotheticalTrue: CellID,
        HypotheticalFalse: CellID,
        Supplied: T,

        pub fn from(value: T) Self {
            return Self { .Supplied=value };
        }

        pub fn eq(self: Self, other: Self) bool {
            switch (self) {
                .HypotheticalTrue => |self_cell_id| {
                    switch (other) {
                        .HypotheticalTrue => |other_cell_id| {
                            return self_cell_id == other_cell_id;
                        },
                        else => return false
                    }
                },
                .HypotheticalFalse => |self_cell_id| {
                    switch (other) {
                        .HypotheticalFalse => |other_cell_id| {
                            return self_cell_id == other_cell_id;
                        },
                        else => return false
                    }
                },
                .Supplied => |self_value| {
                    switch (other) {
                        .Supplied => |other_value| {
                            return @enumToInt(self_value) == @enumToInt(other_value);
                        },
                        else => return false
                    }
                }
            }
        }

        pub fn less_than(self: Self, other: Self) bool {
            switch (self) {
                .HypotheticalTrue => |self_cell_id| {
                    switch (other) {
                        .HypotheticalTrue => |other_cell_id| {
                            return self_cell_id < other_cell_id;
                        },
                        else => return false
                    }
                },
                .HypotheticalFalse => |self_cell_id| {
                    switch (other) {
                        .HypotheticalFalse => |other_cell_id| {
                            return self_cell_id < other_cell_id;
                        },
                        else => return false
                    }
                },
                .Supplied => |self_value| {
                    switch (other) {
                        .Supplied => |other_value| {
                            return @enumToInt(self_value) < @enumToInt(other_value);
                        },
                        else => return false
                    }
                }
            }
        }
    };
}

const Action = union(enum) {
    AmbChoose: CellID
};

pub fn TruthManagementContext(comptime T: type) type {
    const NoGood = struct {
        set: ArrayList(Premise(T))
    };

    return struct {
        const Self = @This();

        premise_outness: ArrayList(Premise(T)),
        premise_nogoods: ArrayList(NoGood),

        pub fn init(allocator: *Allocator) Self {
            return Self {
                .premise_outness=ArrayList(Premise(T)).init(allocator),
                .premise_nogoods=ArrayList(NoGood).init(allocator),
            };
        }

        pub fn is_premise_in(self: *Self, premise: Premise) bool {
            for (self.premise_outness.items) |premise| {
                if (premise == Premise) {
                    return true;
                }
            }

            return false;
        }

        pub fn run_action(self: *Self, action: Action) void {
            //switch (action) {
                //Action.AmbChoose => |cell_id| {
                    //const true_premise = 
                //}
            //}
            
        }
    };
}
